api/pkg/s3/objectcopy.go

```diff
@@ -15,15 +15,204 @@
 package s3
 
 import (
+	"net/url"
+	"strings"
+	"time"
+
 	"github.com/emicklei/go-restful"
-	s3error "github.com/opensds/multi-cloud/s3/error"
+	"github.com/opensds/multi-cloud/api/pkg/common"
+	. "github.com/opensds/multi-cloud/s3/error"
+	"github.com/opensds/multi-cloud/s3/pkg/meta/types"
+	pb "github.com/opensds/multi-cloud/s3/proto"
 	log "github.com/sirupsen/logrus"
 )
 
+func getStorageClassFromHeader(request *restful.Request) (types.StorageClass, error) {
+	storageClassStr := request.HeaderParameter(common.REQUEST_HEADER_STORAGE_CLASS)
+	if storageClassStr != "" {
+		return types.MatchStorageClassIndex(storageClassStr)
+	} else {
+		// If you don't specify this header, Amazon S3 uses STANDARD
+		return types.ObjectStorageClassStandard, nil
+	}
+}
 
-//ObjectPut -
+// ObjectCopy copy object from http header x-amz-copy-source
 func (s *APIService) ObjectCopy(request *restful.Request, response *restful.Response) {
 	log.Infof("received request: Copy object")
-	WriteErrorResponse(response, request, s3error.ErrNotImplemented)
-}
 
+	targetBucketName := request.PathParameter(common.REQUEST_PATH_BUCKET_NAME)
+	targetObjectName := request.PathParameter(common.REQUEST_PATH_OBJECT_KEY)
+	//backendName := request.HeaderParameter(common.REQUEST_HEADER_STORAGE_CLASS)
+	log.Infof("received request: Copy object, objectkey=%s, bucketName=%s\n:",
+		targetObjectName, targetBucketName)
+
+	// copy source is of form: /bucket-name/object-name?versionId=xxxxxx
+	copySource := request.HeaderParameter(common.REQUEST_HEADER_COPY_SOURCE)
+	if copySource == "" {
+		WriteErrorResponse(response, request, ErrInvalidCopySource)
+		return
+	}
+	// Skip the first element if it is '/', split the rest.
+	if strings.HasPrefix(copySource, "/") {
+		copySource = copySource[1:]
+	}
+	splits := strings.SplitN(copySource, "/", 2)
+
+	// Save sourceBucket and sourceObject extracted from url Path.
+	var err error
+	var sourceBucketName, sourceObjectName, sourceVersion string
+	if len(splits) == 2 {
+		sourceBucketName = splits[0]
+		sourceObjectName = splits[1]
+	}
+	// If source object is empty, reply back error.
+	if sourceObjectName == "" {
```

Comment:
Do we need to check if sourceBucketName equals to "" too?

Reply:
I think it is necessary, and how about your suggestion?

Reply:
I agree with you, if sourceBucketName equals to "", that's invalid. How do you think?

Reply:
Thanks for your reviewing, I will add sourceBucketName condition check to it

Reply:
@kumarashit  Thanks,  I add some lines to return error when the parts of len(splits) is not equal 2